### DISCUSSION TITLE
React 18 use mobx without observer

### DISCUSSION THREAD

**Discussion by ikabetskyi-trimac** [mobxjs/mobx#3589] [Ideas]
In React 18 we have useSyncExternalStore hook and we can work with our global store without observer.

const selector = () => mobxStore.data;

const data= useSyncExternalStore(autorun, selector);

**Comment by penx** [comment]
If this works with no caveats, it seems it would be useful to document and export a helper such as

```ts
export const useSelector = <T>(selector: () => T): T => {
 return useSyncExternalStore(autorun, selector);
}
```

([related](https://github.com/mobxjs/mobx/issues/2526#issuecomment-1750522211))

**Comment by urugator** [comment]
Please check the discussion under
https://github.com/mobxjs/mobx/discussions/2566#discussioncomment-557672  (uncollapse the replies)

**Reply by penx** [reply]
ok I think that makes sense, thanks for pointing me to it :)

My understanding is:

- it's safe as long as you're returning primitives and not observables
- you don't want this as an export from mobx-react, due to the maintenance burden

with that in mind, I _think_ this should be safe to use

```ts
export const useSelector = <T>(selector: () => T): T => {
  const checkSelector = useCallback(() => {
    const value = selector();
    if (isObservable(value)) {
      throw new Error("useSelector cannot be used on an observable");
    }
    return value;
  }, [selector]);
  return useSyncExternalStore(autorun, checkSelector);
};
```
